lib/Driver/Compilation.cpp

```diff
@@ -589,7 +678,137 @@ namespace driver {
       }
     }
 
+    /// Insert all jobs in \p Cmds (of descriptive name \p Kind) to the \c
+    /// TaskQueue, and clear \p Cmds.
+    void transferJobsToTaskQueue(CommandSet &Cmds, StringRef Kind) {
+      for (const Job *Cmd : Cmds) {
+        if (Comp.ShowJobLifecycle)
+          llvm::outs() << "Adding " << Kind
+                       << " job to task queue: "
+                       << LogJob(Cmd) << "\n";
+        addPendingJobToTaskQueue(Cmd);
+      }
+      Cmds.clear();
+    }
+
+    /// Partition the jobs in \c PendingExecution into those that are \p
+    /// Batchable and those that are \p NonBatchable, clearing \p
+    /// PendingExecution.
+    void getPendingBatchableJobs(CommandSet &Batchable,
+                                 CommandSet &NonBatchable) {
+      for (const Job *Cmd : PendingExecution) {
+        if (Comp.getToolChain().jobIsBatchable(Comp, Cmd)) {
+          if (Comp.ShowJobLifecycle)
+            llvm::outs() << "Batchable: " << LogJob(Cmd) << "\n";
+          Batchable.insert(Cmd);
+        } else {
+          if (Comp.ShowJobLifecycle)
+            llvm::outs() << "Not batchable: " << LogJob(Cmd) << "\n";
+          NonBatchable.insert(Cmd);
+        }
+      }
+      PendingExecution.clear();
+    }
+
+    /// If \p CurrentBatch is nonempty, construct a new \c BatchJob from its
+    /// contents by calling \p ToolChain::constructBatchJob, then insert the
+    /// new \c BatchJob into \p Batches and clear \p CurrentBatch.
+    void
+    formBatchJobFromCurrentBatch(CommandSet &Batches,
+                                 llvm::SetVector<const Job *> &CurrentBatch) {
+      if (CurrentBatch.empty())
+        return;
+      if (Comp.ShowJobLifecycle)
+        llvm::outs() << "Forming batch job from "
+                     << CurrentBatch.size() << " constituents\n";
+      auto const &TC = Comp.getToolChain();
+      auto J = TC.constructBatchJob(CurrentBatch.getArrayRef(), Comp);
+      if (J)
+        Batches.insert(Comp.addJob(std::move(J)));
+      CurrentBatch.clear();
+    }
+
+    /// Return true iff \p Cmd can be expanded by \p CurrentBatch, meaning
+    /// that \p CurrentBatch is smaller than \p TargetBatchSize and \p Cmd
+    /// is batch-combinable with the equivalence class of \p CurrentBatch
+    /// (as represented by element 0 of \p CurrentBatch).
+    bool canExpandBatch(const Job *Cmd,
+                        llvm::SetVector<const Job *> &CurrentBatch,
+                        size_t TargetBatchSize) {
+      if (CurrentBatch.empty())
+        return true;
+      auto const &TC = Comp.getToolChain();
+      if (TC.jobsAreBatchCombinable(Comp, Cmd, CurrentBatch[0]))
+        return true;
+      return CurrentBatch.size() < TargetBatchSize;
```

Comment:
It looks like these conditions are backwards: you're doing an OR rather than an AND check here.

Reply:
Ah good catch.